tools/python/mwm_downloader.py

```diff
@@ -0,0 +1,145 @@
+
+
+def download_file(url, filename, attempts=3):
+    try:
+        urllib.request.urlretrieve(url, filename=filename)
+    except urllib.error.HTTPError as err:
+        if 400 <= err.code < 500:
+            logger.error('URL \'{url}\' is not found'.format(url=url), exc_info=True)
+            return False
+    except PermissionError:
+        logger.error('Can\'t write file {filename}: permission denied'.format(filename=filename), exc_info=True)
+        return False
+    except urllib.error.URLError:
+        if attempts:
+            time.sleep(3)
+            return download_file(url, filename, attempts - 1)
+        logger.error('File {filename} is not loaded'.format(filename=filename), exc_info=True)
+        return False
+    return True
+
+
+def progress_bar(total, progress):
+    """
+    Displays or updates a console progress bar.
+
+    Original source: https://stackoverflow.com/a/15860757/1391441
+    """
+    initial_progress = progress
+    bar_length = 20
+    progress = progress / total
+    block = int(round(bar_length * progress))
+    progress_bar_str = '#' * block + '-' * (bar_length - block)
+    text = '\r[{progress_bar}] {percent:.0f}% {item}/{total}'.format(
+        progress_bar=progress_bar_str,
+        percent=round(progress * 100, 0),
+        item=initial_progress,
+        total=total)
+    sys.stderr.write(text)
+    if progress == 1:
+        sys.stderr.write('\n')
+    sys.stderr.flush()
+
+
+def download_mwm_list(version=None, threads=8, folder=None, mwm_prefix_list=None, quiet=True):
+    versions = list(get_mwm_versions())
+
+    if version is None:
+        # Get latest version by default
+        version = max(versions)
+    elif version not in versions:
+        logger.error('Maps with version {version} is not found in {url}'.format(version=version, url=DIRECT_MAP_URL))
+        return
+
+    mwm_names = get_mwm_names(version)
+    if mwm_prefix_list is not None:
+        mwm_regexp = re.compile('|'.join(mwm_prefix_list))
+        mwm_to_download = filter(lambda m: mwm_regexp.match(m), mwm_names)
+    else:
+        mwm_to_download = mwm_names
+
+    if folder is None:
+        folder = '{base}/{version}'.format(base=DEFAULT_DOWNLOAD_DIRECTORY, version=version)
+
+    try:
+        os.makedirs(folder)
+    except OSError:
+        ...
+
+    with ThreadPoolExecutor(max_workers=threads) as executor:
+        tasks = []
+        for mwm_name in mwm_to_download:
+            unescaped_name = urllib.parse.quote(mwm_name)
+            url = '{base}{version}/{mwm_name}'.format(base=DIRECT_MAP_URL, version=version, mwm_name=unescaped_name)
+            filename = '{folder}/{mwm_name}'.format(folder=folder, mwm_name=mwm_name)
+            tasks.append(executor.submit(download_file, url, filename))
+
+        # No sense to show progress bar for one map
+        if len(tasks) < 2:
+            quiet = True
+
+        for num, _ in enumerate(as_completed(tasks)):
```

Comment:
What if task failed? Don't you want to say something to the user at the end?

Reply:
We log exceptions inside `download_file`, user will see messages in stderr, i think it's enough in this case.